### PR TITLE
Reduce allocations and retained memory in uri/common.rb

### PR DESCRIPTION
**Pull request by SamSaffron**
Reduce allocations in uri/common.rb

Opening a ticket on redmine to discuss

### CODE REVIEW

File: lib/uri/common.rb [hunk 1]
```diff
@@ -342,17 +342,17 @@ def self.regexp(schemes = nil)
 
   TBLENCWWWCOMP_ = {} # :nodoc:
   256.times do |i|
-    TBLENCWWWCOMP_[i.chr] = '%%%02X' % i
+    TBLENCWWWCOMP_[-i.chr] = -('%%%02X' % i)
   end
-  TBLENCWWWCOMP_[' '] = '+'
+  TBLENCWWWCOMP_[' '] = -'+'
```
Comment (k0kubun) [inline lib/uri/common.rb]: Why do you call `String#-@` in addition to specifying `# frozen_string_literal: true`? The `'+'` literal should be frozen and does no allocation without calling it. All other `String#-@` calls seem lengthy too.

**Comment by k0kubun** [comment]
The failing test should be fixed.

**Comment by SamSaffron** [comment]
OK, made a ticket in redmine here: https://bugs.ruby-lang.org/issues/14410

I think I corrected both issues @k0kubun

**Comment by hsbt** [comment]
merged at 26b3dc49226011469bfe29af0f882b06215388a5

**Pull request closed without merging**